velox/exec/PrefixSort.cpp

```diff
@@ -0,0 +1,320 @@
+        prefixSortLayout.compareFlags[index].ascending,
+        *(reinterpret_cast<nativeType*>(row + rowColumn.offset())),
+        prefix + prefixSortLayout.prefixOffsets[index]);
+  }
+}
+
+template <>
+FOLLY_ALWAYS_INLINE void normalizeRowColumn<TypeKind::BIGINT>(
+    const PrefixSortLayout& prefixSortLayout,
+    const uint32_t index,
+    const RowColumn& rowColumn,
+    char* const row,
+    char* const prefix) {
+  normalizeNumericColumn<
+      TypeKind::BIGINT,
+      typename TypeTraits<TypeKind::BIGINT>::NativeType>(
+      prefixSortLayout, index, rowColumn, row, prefix);
+}
+
+template <>
+FOLLY_ALWAYS_INLINE void normalizeRowColumn<TypeKind::INTEGER>(
+    const PrefixSortLayout& prefixSortLayout,
+    const uint32_t index,
+    const RowColumn& rowColumn,
+    char* const row,
+    char* const prefix) {
+  normalizeNumericColumn<
+      TypeKind::INTEGER,
+      typename TypeTraits<TypeKind::INTEGER>::NativeType>(
+      prefixSortLayout, index, rowColumn, row, prefix);
+}
+} // namespace
+
+PrefixSortLayout PrefixSortLayout::makeSortLayout(
+    const std::vector<TypePtr>& types,
+    const std::vector<CompareFlags>& keyCompareFlags,
+    const uint32_t maxNormalizedKeySize) {
+  VELOX_CHECK_GE(types.size(), 0);
+  uint32_t normalizedKeySize = 0;
+  uint32_t numNormalizedKeys = 0;
+  uint32_t numSortKeys = types.size();
+  std::vector<uint32_t> prefixOffsets;
+  for (const auto& type : types) {
+    if (normalizedKeySize > maxNormalizedKeySize) {
+      break;
+    }
+    if (type->kind() == TypeKind::BIGINT) {
+      prefixOffsets.push_back(numNormalizedKeys++);
+      normalizedKeySize += sizeof(TypeTraits<TypeKind::BIGINT>::NativeType);
+    } else if (type->kind() == TypeKind::INTEGER) {
+      prefixOffsets.push_back(numNormalizedKeys++);
+      normalizedKeySize += sizeof(TypeTraits<TypeKind::INTEGER>::NativeType);
+    } else {
+      break;
+    }
+  }
+  uint32_t compareMask = normalizedKeySize % 8;
+  uint32_t remainIntOffset = (normalizedKeySize / 8) * 8;
+  uint32_t remainShortOffset = remainIntOffset + (compareMask & 0b100) * 4;
+  uint32_t remainByteOffset = remainShortOffset + (compareMask & 0b010) * 2;
+
+  return PrefixSortLayout{
+      normalizedKeySize + sizeof(char*),
+      normalizedKeySize,
+      numNormalizedKeys,
+      numSortKeys,
+      keyCompareFlags,
+      numNormalizedKeys < numSortKeys,
+      std::move(prefixOffsets),
+      compareMask,
+      remainIntOffset,
+      remainShortOffset,
+      remainByteOffset};
+}
+
+// For normalizedKeySize > 8, use uint64_t word compare.
+// For remain part of the binary string, the size has only 7 cases, and
+// any number less than 8 can be represented by 3 bits:
+// highest bit means 4, we use integer compare,
+// middle bit means 2, we use memcmp(char*, char*, 2)
+// lowest bit means 1, we use memcmp(char*, char*, 1)
+FOLLY_ALWAYS_INLINE int
+PrefixSort::memCompareByWord(char* left, char* right, int size) {
+  for (int i = 0; i < size / 8; ++i) {
```

Comment:
Is this somehow faster than std::memcmp ? We may need a separate benchmark to demonstrate that.

Reply:
The opt logic is that:
1. reduce loop times
2. value compare helps apply auto-vectorized / simd

benchmark result :
use memCompareByWord:
![image](https://github.com/facebookincubator/velox/assets/13514695/43ae4761-dcb0-4e54-b0e6-41cdcf9fc14b)

use std::memcmp:
![image](https://github.com/facebookincubator/velox/assets/13514695/b41e7d49-3c1a-4120-bf57-72d227218d07)